### ISSUE TITLE
path/filepath: COM0 and LPT0 are now reserved on Windows

### ISSUE THREAD

**Issue by neild** [golang/go#67245]
https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file states that COM0 and LPT0 are reserved names. I believe that this page used to only list COM[1-9] and LPT[1-9] as reserved, and empirically these names are available for use on the Windows versions I've tested.

We need to decide whether `filepath.IsLocal` should consider these names local or not.

In favor of `filepath.IsLocal("COM0")==false`: Current behavior, empirically you can create a file named `COM0`.

In favor of `filepath.IsLocal("COM0")==true`: Documented as a reserved name.

**Comment by ianlancetaylor** [comment]
CC @golang/windows

**Comment by bjorndm** [comment]
https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#win32-file-namespaces

It seems very complex. Since these are aliases, on a computer without serial or parallel ports these file names can be created, but it is not portable.

**Comment by neild** [comment]
Huh. I wonder if the reference to COM0 is a typo in https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#win32-file-namespaces.

I don't understand how any of this works under the hood, but there's some form of special consideration given to the names `COM[1-2¹²³]`, regardless of whether there's an associated serial port, but not for `COM0`.

**Comment by bjorndm** [comment]
Maybe someone could test with a few USB to serial adapters and see what COM number gets assigned? Or even better , on a desktop with a serial PCI extension board? If COM0 never gets assigned in practice, we can keep this as it is now.

**Comment by alexbrainman** [comment]
> https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file states that COM0 and LPT0 are reserved names. I believe that this page used to only list COM[1-9] and LPT[1-9] as reserved, and empirically these names are available for use on the Windows versions I've tested.

Agree. This matches my experience. `COM[1-9]` and `LPT[1-9] ` are reserved names on Windows, and should only be used to refer to existing serial or parallel ports.

> Huh. I wonder if the reference to COM0 is a typo in https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#win32-file-namespaces.

I don't see `COM0` mentioned in https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#win32-file-namespaces section.

> I don't understand how any of this works under the hood, ...

Same here. Perhaps other Windows developers know more.

Alex

**Comment by qmuntal** [comment]
> I believe that this page used to only list COM[1-9] and LPT[1-9] as reserved

True, LPT0 and COM0 were added two years ago (by a non-Microsoft contributor): https://github.com/MicrosoftDocs/win32/commit/2274ec1c189564d9fbb6f4a6d885ea6cc95f059a.

A more recent commit in that same file suggests that those names might be valid file names: https://github.com/MicrosoftDocs/win32/commit/b4c2d78eb8c7cfa9d3dc1a94d2064a9078445125

> Tangentially, COM0 and LPT0 are in practice treated as normal file names, contrary to documentation; I have not changed this as it's not clear whether they're still reserved for future use or just in case.

I'm on leave now, so can't contact the owners to know the right answer.

**Comment by matloob** [comment]
That's really strange. We might want to undo the change to x/mod to exclude those names until we get better information. It's pretty concerning that the change was made by a non-Microsoft contributor but doesn't have more information in the commit message.

**Comment by neild** [comment]
I think for `path/filepath`, we should stick with treating LPT0 and COM0 as non-reserved. Whether they're officially "reserved" or not, they're empirically available for use and I can't imagine Microsoft with their admirable commitment to backwards compatibility changing that at this point. All this behavior has been stable for decades.

I don't have an opinion on x/mod.

**Comment by qmuntal** [comment]
The docs have been updated. COM0 and LPT0 are not reserved: https://github.com/MicrosoftDocs/win32/commit/4fde3c1de423dece99b0e077bd651b9aa327c0bf.
